README.md
SleekGallery
==========
A [jQuery] [1] port of [JonDesign's SmoothGallery] [2]

###What is it?
Using jQuery 1.3.2 or newer, this javascript gallery and slideshow system allows you to have simple and smooth (cross-fading...) image galleries, slideshows, showcases and other cool stuff on your website... 

###So, what is so cool about it ? 
Unlike other systems out there, JonDesign's SmoothGallery was designed from the ground up to be standard compliant: You can feed it from any document, using custom css selectors.

SleekGallery continues in this spirit and tries to be flexible, configurable and yet as lightweight as possible. Currently the base slideshow script is 37kb uncompressed. 

This can be reduced to less than 7kb using javascript minification and gzip compression on your webserver.

###Why consider it instead of a Flash-based solution ?
* Let's imagine you want to add a showcase of your last products, or even a showcase of one product on your homepage. Would you want to restrict it only to the users who have Flash enabled ?
* Another important point is the fact that using this script, you will have fully standard compliant web pages (which is important for accessibility, for example).
* Using a flash based solution, search engines won't see your content nor links. Not really the expected result, right ?
* Oh, and do you really want to rely on a big company's proprietary licensed app for your website ?

###On what browsers does it run ? 
Since it's based on the jQuery library and the dynamically generated code is fully (X)HTML compliant it should support all all modern and capable browsers.

Only browsers on the current YUI Graded Browser Support ["A-Grade"] [3] chart will officially be supported regarding visual features.

###What's currently missing compared to SmoothGallery 2.1beta1
* HistoryManager so that the browsers back and forward buttons control the gallery
* The "enhanced" set of transitions has not been ported yet
* Gallery Sets have not been ported yet

  [1]: http://jquery.com/ "jQuery"
  [2]: http://smoothgallery.jondesign.net/ "JonDesign's SmoothGallery"
  [3]: http://developer.yahoo.com/yui/articles/gbs/#gbschart "A-Grade"


Changelog
v1.1.1:
  * Fix preloading of images when using Internet Explorer 6/7/8
  * Fix slideInfoZone animating twice when selecting a new gallery set
  * Fix syntax in pass() definition to make IE happy

v1.1.0:
  * Implement gallery sets as available in SmoothGallery 2.1beta

v1.0.5:
  * Fix flushGallery() and flushThumbnails()
  * Implement custom event triggers
  * Whitespace cleanup

v1.0.4:
  * Fixed bug with FancyBox not being initalized on first image
  * Fixed bug with FancyBox not preventing the default click event
  * Added support for image groups in FancyBox

v1.0.3:
  * Ported additional transitions "fadeslideleft", "continuoushorizontal" and "continuosvertical"
  * Added demos for new transitions

v1.0.2:
  * Implemented transitions "fadebg" and "crossfade"

v1.0.1:
  * Implemented support for FancyBox to show big images

v1.0:
  * Fork from JonDesign's SmoothGallery - Initial release based on jQuery


scripts/jquery.sleekgallery.transitions.js
/*
 * This file is part of SleekGallery v1.1.0.
 *
 * SleekGallery is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * SleekGallery is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoJo's SlickGallery; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Main Developer: Morton Jonuschat <mjonuschat@gmail.com> - http://github.com/yabawock
 */
(function($) {
    $.extend(true, {
        sleekgallery: {
            Transitions: {
                fadeslideleft: function(oldImage, newImage, oldPos, newPos) {
                    var duration = 1500;
                    var transition = 'swing';
                    if(jQuery.isFunction(jQuery.easing.easeOutCubic)) {
                        transition = 'easeOutCubic';
                    }
                    if (newPos > oldPos) {
                        newImage.css({'left' : this.galleryElement.outerWidth(true)}).animate({'left' : 0, 'opacity' : 1}, duration, transition);
                        oldImage.css({'opacity' : 1}).animate({'opacity' : 0}, duration, transition);
                    } else {
                        newImage.css({'opacity' : 0}).animate({'opacity' : 1}, duration, transition);
                        oldImage.css({'left' : 0}).animate({'left' : this.galleryElement.outerWidth(true), 'opacity' : 0}, duration, transition, function() { this.css('left', 0); });
                    }
                },
                continuoushorizontal: function(oldImage, newImage, oldPos, newPos) {
                    if (((newPos > oldPos) || ((newPos==0) && (oldPos == (this.maxIter-1) ))) &&
                        (!((newPos == (this.maxIter-1 )) && (oldPos == 0)))) {
                        oldImage.css({'opacity' : 1, 'left' : 0}).animate({
                            'left' : this.galleryElement.outerWidth(true) * -1
                            }, this.options.fadeDuration, 'linear');
                        newImage.css({'opacity' : 1, 'left' : this.galleryElement.outerWidth(true)}).animate({
                            'left' : 0
                        }, this.options.fadeDuration, 'linear');
                    } else  {
                        oldImage.css({'opacity' : 1, 'left' : 0}).animate({
                            'left' : this.galleryElement.outerWidth(true)
                        }, this.options.fadeDuration, 'linear');
                        newImage.css({'opacity' : 1, 'left': this.galleryElement.outerWidth(true) * -1}).animate({
                            'left' : 0
                        }, this.options.fadeDuration, 'linear');
                    }
                },
                continuousvertical: function(oldImage, newImage, oldPos, newPos) {
                    if (((newPos > oldPos) || ((newPos==0) && (oldPos == (this.maxIter-1) ))) &&
                        (!((newPos == (this.maxIter-1 )) && (oldPos == 0)))) {
                        oldImage.css({'opacity' : 1, 'top' : 0}).animate({
                            'top' : this.galleryElement.outerHeight(true) * -1
                        }, this.options.fadeDuration, 'linear');
                        newImage.css({'opacity' : 1, 'top' : this.galleryElement.outerHeight(true)}).animate({
                            'top' : 0
                        }, this.options.fadeDuration, 'linear');
                    } else {
                        oldImage.css({'opacity' : 1, 'top' : 0}).animate({
                            'top' : this.galleryElement.outerHeight(true)
                        }, this.options.fadeDuration, 'linear');
                        newImage.css({'opacity' : 1, 'top' : this.galleryElement.outerHeight(true) * -1}).animate({
                            'top' : 0
                        }, this.options.fadeDuration, 'linear');
                    }
                }
            }
        }
    });
})(jQuery);


scripts/jquery.sleekgallery.set.js
/*
 * This file is part of SleekGallery v1.1.0.
 *
 * SleekGallery is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * SleekGallery is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoJo's SlickGallery; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *
 * Main Developer: Morton Jonuschat <mjonuschat@gmail.com> - http://github.com/yabawock
 */
(function($) {
    Function.prototype.pass = function () {
        var instance = this, length = arguments.length, args = new Array(), object;
        while(length--) {
            args[length] = arguments[length];
        }
        object = args.shift();
        return function() {
            return instance.apply(object, args);
        };
    };

    // Plugin definition
    $.fn.extend({
        sleekGallerySet: function(options) {
            return this.each(function() {
                jQuery.sleekgallerySet.initializeSet(this, options);
            });
        }
    });

    $.extend(true, {
        sleekgallerySet: jQuery.sleekgallery
    });

    // Extend sleekGallery with Group/Set functions
    $.extend(true, {
        sleekgallerySet: {
            setOptions: {
                manualSetData: [],
                gallerySelector: "div.galleryElement",
                galleryTitleSelector: "h2",
                textGallerySelector: 'Galleries',
                textShowGallerySelector: 'Other Galleries',
                textGalleryInfo: '{0} pictures',
                startWithSelector: true,
                activateSelectorScroller: true,
                /* Changing default options */
                textShowCarousel: '{0}/{1} Pictures',
                carouselPreloader: false
            },
            initializeSet: function(element, options) {
                this.options = jQuery.extend({}, this.options, this.setOptions, options);
                if(!jQuery.isFunction(jQuery.scrollTo)) {
                    this.options.activateSelectorScroller = false;
                }
                this.gallerySet = this.options.manualSetData;
                jQuery(this).bind('onPopulated', this.createGallerySelectorTab.pass(this));
                jQuery(this).bind('onPopulated', this.createGallerySelector.pass(this));
                this.startWithSelectorFn = this.toggleGallerySelector.pass(this, true);
                if (this.options.startWithSelector) {
                    jQuery(this).bind('onGallerySelectorCreated', this.startWithSelectorFn);
                }
                this.initialize(element, this.options);
            },
            populateData: function() {
                options = this.options;
                var data = jQuery.makeArray(this.gallerySet);
                jQuery(options.gallerySelector).each(function (index, galEl) {
                    currentGalArrayPlace = 0;
                    galleryDict = {
                        title: jQuery(options.galleryTitleSelector, this).html(),
                        elements: jQuery(galEl)
                    };
                    jQuery.merge(data, [galleryDict]);
                });
                for(i=0;i<data.length;i++) {
                    galEl = data[i].elements;
                    data[i].elements = this.populateGallery(galEl, 0);
                    if (this.options.destroyAfterPopulate) {
                        galEl.remove();
                    }
                }
                this.gallerySet = data;
                this.galleryData = data[0].elements;
                this.currentGallery = 0;
                jQuery(this).trigger('onPopulated');
            },
            changeGallery: function(number)
            {
                if (number!=this.currentGallery)
                {
                    this.changeData(this.gallerySet[number].elements);
                    this.maxIter = this.gallerySet[number].elements.length;
                    this.currentGallery = number;
                    jQuery(this.gallerySelectorBtn).html(this.gallerySet[number].title);
                    jQuery(this).trigger('onGalleryChanged');
                }
                this.toggleGallerySelector(false);
            },
            createGallerySelectorTab: function() {
                this.gallerySelectorBtn = jQuery('<a>').addClass('gallerySelectorBtn').css({
                    'title': this.options.textShowGallerySelector
                }).html(this.options.textShowGallerySelector).click(function(){ this.toggleGallerySelector(true); }.pass(this)
                ).appendTo(this.galleryElement);
                jQuery(this).bind('onShowCarousel', function() {
                    this.gallerySelectorBtn.css({'zIndex': 10});
                }.pass(this));
                jQuery(this).bind('onCarouselHidden', function(){
                    this.gallerySelectorBtn.css({'zIndex': 15});
                }.pass(this));
            },
            createGallerySelector: function() {
                this.gallerySelector = jQuery('<div>').addClass(
                        'gallerySelector'
                    ).appendTo(
                        this.galleryElement
                    ).css({
                        'display': 'none',
                        'opacity': '0'
                    });

                this.gallerySelectorTitle = jQuery('<h2>').html( this.options.textGallerySelector).appendTo(this.gallerySelector);
                var gallerySelectorHeight = this.galleryElement.outerHeight(true) - 50 - 10 - 2;
                this.gallerySelectorWrapper = jQuery('<div>').addClass(
                        'gallerySelectorWrapper'
                    ).css(
                        'height',
                        gallerySelectorHeight + "px"
                    ).appendTo(this.gallerySelector);
                this.gallerySelectorInner = jQuery('<div>').addClass('gallerySelectorInner').appendTo(this.gallerySelectorWrapper);
                this.createGalleryButtons();
                jQuery(this).trigger('onGallerySelectorCreated');
            },
            createGalleryButtonHelper: function(index, galleryItem) {
                var galleryButtonWidth =
                    ((this.galleryElement.outerWidth(true) - 30) / 2) - 14;
                var button = jQuery('<div>').addClass('galleryButton').appendTo(
                    this.gallerySelectorInner
                ).bind('mouseenter', function(myself){
                        myself.button.addClass('hover');
                    }.pass(this, galleryItem)
                ).bind('mouseleave', function(myself){
                        myself.button.removeClass('hover');
                    }.pass(this,galleryItem)
                ).bind('click', function(myself, number){
                        this.changeGallery.pass(this,number)();
                    }.pass(this, galleryItem, index)
                ).css('width', galleryButtonWidth);
                galleryItem.button = button;
                var thumbnail = "";
                if (this.options.showCarousel) {
                    thumbnail = galleryItem.elements[0].thumbnail;
                } else {
                    thumbnail = galleryItem.elements[0].image;
                }
                jQuery('<div>').addClass('preview').css(
                    'backgroundImage',
                    "url('" + thumbnail + "')"
                ).appendTo(button);
                jQuery('<h3>').html(galleryItem.title).appendTo(button);
                jQuery('<p>').addClass('info').html(this.printf(this.options.textGalleryInfo, galleryItem.elements.length)).appendTo(button);
            },
            createGalleryButtons: function () {
                for(i=0;i<this.gallerySet.length;i++) { this.createGalleryButtonHelper(i, this.gallerySet[i]); }
                jQuery('<br>').appendTo(this.gallerySelectorInner).css('clear','both');
            },
            toggleGallerySelector: function(state) {
                if (state) {
                    jQuery(this.gallerySelector).css('display','block').animate({'opacity' : 1}, this.options.fadeDuration, function(){
                        this.gallerySelectorWrapper.unbind('mousemove').bind('mousemove', this, this.selectorScroll);
                        this.gallerySelectorWrapper.unbind('mouseleave').bind('mouseleave', this, function() {
                            this.gallerySelectorWrapper.stop();
                            this.gallerySelectorInner.scroll = null;
                        }.pass(this));
                    }.pass(this));
                } else {
                    jQuery(this.gallerySelector).animate({'opacity' : 0}, this.options.fadeDuration, 'linear', function() {
                        this.gallerySelectorWrapper.stop();
                        this.gallerySelectorInner.scroll = null;
                        this.gallerySelectorWrapper.unbind('mousemove');
                        this.gallerySelectorWrapper.unbind('mouseleave');
                        jQuery(this.gallerySelector).css({'display':'none'});
                    }.pass(this));
                }
            },
            selectorScroll: function(e) {
                var y = e.pageY - e.data.gallerySelectorWrapper.offset().top;
                if(y <= (e.data.gallerySelectorWrapper.outerHeight(true)*.2)) {
                    if(e.data.gallerySelectorInner.scroll != 'top') {
                        e.data.gallerySelectorInner.scroll = 'top';
                        var duration = (e.data.options.carouselVelocity * 5) * e.data.gallerySelectorWrapper.scrollTop();
                        e.data.gallerySelectorWrapper.scrollTo(0,duration);
                    }
                } else if(y >= (e.data.gallerySelectorWrapper.outerHeight(true) - (e.data.gallerySelectorWrapper.outerHeight(true)*.2))) {
                    if(e.data.gallerySelectorInner.scroll != 'bottom') {
                        e.data.gallerySelectorInner.scroll = 'bottom';
                        var duration = (e.data.options.carouselVelocity * 5) * (e.data.gallerySelectorInner.outerHeight(true) - e.data.gallerySelectorWrapper.outerHeight(true) - e.data.gallerySelectorWrapper.scrollTop());
                        e.data.gallerySelectorWrapper.scrollTo('max',duration);
                    }
                } else {
                    if(e.data.gallerySelectorInner.scroll) {
                        e.data.gallerySelectorInner.scroll = null;
                        e.data.gallerySelectorWrapper.stop();
                    }
                }
            },
            initHistory: function() {
                // TODO: Find a jQuery history plugin
            }
        }
    });
})(jQuery);
